[WIP] Atomic bsq trades

sqrrm opened the pull request:
<!-- 
- make yourself familiar with the CONTRIBUTING.md if you have not already (https://github.com/bisq-network/bisq/blob/master/CONTRIBUTING.md)
- make sure you follow our [coding style guidelines][https://github.com/bisq-network/style/issues)
- pick a descriptive title
- provide some meaningful PR description below
- create the PR
- in case you receive a "Change request" and/or a NACK, please react within 30 days. If not, we will close your PR and it can not be up for compensation.
- After addressing the change request, __please re-request a review!__ Otherwise we might miss your PR as we tend to only look at pull requests tagged with a "review required".
-->

Replaces https://github.com/bisq-network/bisq/pull/5121/, https://github.com/bisq-network/bisq/pull/5127

Major refactoring of OfferPayload to accommodate offers with no maker fee tx.

A lot of copy paste and remove extra stuff to place and take atomic offers.

Trade domain refactoring to allow for atomic trades.

sqrrm committed:
Add atomic account type

Add a hidden account type that's automatically used for BSQ trades. This
means old orders not using AtomicAccount can still complete the trades
as per normal and new offers placed with AtomicAccount can be taken by
anyone with an upgraded client, but won't be possible to take by users
with older clients.

The atomic account is added on startup if not already added. There is no
data associated with the account, a new BSQ address will be chosen
automatically during the atomic trade process.

sqrrm committed:
Add UI for creating atomic offers

sqrrm committed:
Refactor direct offerPayload access

Broaden type requirement

sqrrm committed:
Refactor direct offerPayload access

Let Offer handle all OfferPayload data access.

sqrrm committed:
Add OfferPayloadI

An interface for OfferPayload currently only used by OfferPayload. Some
features won't be available for newer payload types but to fit this into
the Offer and Trade framework it's necessary to hack it a bit.

sqrrm committed:
Add AtomicOfferPayload

Offer protobuf has oneof FeeTxOfferPayload and AtomicOfferPayload. This is
backwards compatible since the wire format is the same for the old Offer
and the new one with offer_payload set since the index is the same.

sqrrm committed:
Refactor: move OfferPayload Direction protobuf methods

sqrrm committed:
Add UI for taking atomic offers

Copy paste of normal take offer code with extra steps removed

Added AtomicTxBuilder prototype as a way to build the atomic tx, verify
inputs and gather input requirements. Only the input requirement is used
during the take offer process.

sqrrm committed:
Refactor: TradeManager handle Tradable

This prepares for letting TradeManager handle AtomicTrade that is a Tradable

sqrrm committed:
Refactor: add class tradeModel

sqrrm committed:
Introduce interface ProcessModelI

sqrrm committed:
Compiling old atomic trade protocol

sqrrm committed:
Refactor: TradeManager use TradeModel

sqrrm committed:
Refactor: extract applyFilter method

sqrrm committed:
Add AtomicApplyFilter

sqrrm committed:
TradeTaskRunner handle AtomicTrade

sqrrm committed:
Refactor: move AtomicApplyFilter

sqrrm committed:
Take atomic offer

sqrrm committed:
Refactor: rename txFee -> txFeePerVbyte

sqrrm committed:
Add atomic trade protocol

Add atomic trade tab under portfolio

sqrrm committed:
Set maker fee type in offer

sqrrm committed:
Display Atomic Trade info in transaction lists

Set txid when parsing AtomicTrade from protobuf

sqrrm committed:
Add ProofOfWorkPayload interface

sqrrm committed:
Refactor: extract AtomicTakeOfferModel to core

sqrrm committed:
Add ProofOfWorkPayload filter

sqrrm committed:
Remove Arbitrator/Mediator requirement for atomic trades

ghubstan committed:
Add atomic offer/trade grpc proto defs

- Add offers service protos: GetAtomicOffer, GetMyAtomicOffer,
  GetAtomicOffers, GetMyAtomicOffers, CreateAtomicOffer.

- Add proto message AtomicOfferInfo.

- Add bool tradeAtomic field to CreateCryptoCurrencyPaymentAccountRequest.

- Add trades service protos: GetAtomicTrade, TakeAtomicOffer.

- Add proto message AtomicTradeInfo.

ghubstan committed:
Add grpc boilerplate for atomic offer/trade support

ghubstan committed:
Add atomic offer/trade proto wrappers

ghubstan committed:
Add Optional<AtomicTrade> getAtomicTradeById(String tradeId)

ghubstan committed:
Add @Getter annotation to private AtomicTxBuilder atomicTxBuilder

ghubstan committed:
Add boolean isAtomicOffer() convenience

ghubstan committed:
Implement core api's atomic offer/trade support

ghubstan committed:
Support atomic payment method

ghubstan committed:
Add atomic offer/trade tests

sqrrm committed:
Use segwit validator for BSQ output address

sqrrm committed:
Extract fee handling from AtomicTxBuilder

sqrrm committed:
Core: remove BTC trade fee option for atomic trades

All trade fees are now paid using BSQ for atomic trades. The UI elements
have still not been removed

sqrrm committed:
Desktop: remove trade fee option from UI

Clean up some copy paste residue from UI classes

Remove unused AtomicModel

sqrrm committed:
Refactor: OpenOfferManager centralize add/remove

sqrrm committed:
Atomic funding

sqrrm committed:
AtomicOfferPayload: add hash

sqrrm committed:
bitcoinj: update to better multi wallet handling

Use bitcoinj version that clones incoming transactions.

This fixes inconsistent wallet problems caused by bisq creating
transactions with outputs going to both the BSQ and BTC wallets.

ripcurlx commented on core/src/main/java/bisq/core/account/witness/AccountAgeWitnessService.java:
```diff
@@ -425,7 +426,7 @@ private long getTradeLimit(Coin maxTradeLimit,
         if (CurrencyUtil.isCryptoCurrency(currencyCode) ||
                 !PaymentMethod.hasChargebackRisk(paymentMethod, currencyCode) ||
-                direction == OfferPayload.Direction.SELL) {
+                direction == OfferPayloadI.Direction.SELL) {
```
The naming of the interface `OfferPayloadI` somehow feels weird. How about calling the interface `OfferPayload` and the implementation `OfferPayloadImpl` and coding as much as possible against the interface? Or call the implementation something other more specific (I'm to early in the commits to suggest something better - `OfferWithFeePayload`?)

ripcurlx replied:
Or is this about the proto resolving to be backwards compatible?

ripcurlx commented on core/src/main/java/bisq/core/offer/AtomicOfferPayload.java:
```diff
@@ -0,0 +1,207 @@
+
+    @Override
+    public String getBaseCurrencyCode() {
+        return "BSQ";
```
as this is an BSQ atomic offer payload shall we call it AtomicBsqOfferPayload in case we want to support more Atomic offers in the future (XMR?)

ripcurlx commented on core/src/main/java/bisq/core/offer/OpenOfferManager.java:
```diff
@@ -409,6 +412,35 @@ public void placeOffer(Offer offer,
+                user,
+                filterManager);
+
+        var atomicPlaceOfferProtocol = new AtomicPlaceOfferProtocol(atomicModel,
```
Maybe `PlaceAtomicOfferProtocol` instead of `AtomicPlaceOfferProtocol`?

ripcurlx commented on core/src/main/java/bisq/core/offer/placeoffer/atomic/AtomicPlaceOfferModel.java:
```diff
@@ -0,0 +1,70 @@
+
+@Slf4j
+@Getter
+public class AtomicPlaceOfferModel implements Model {
```
Same here `PlaceAtomicOfferModel`?

ripcurlx commented on desktop/src/main/java/bisq/desktop/main/offer/AtomicOfferDataModel.java:
```diff
@@ -0,0 +1,591 @@
+        Volume volumeByAmount = price.get().getVolumeByAmount(minAmount.get());
+
+        // For HalCash we want multiple of 10 EUR
+        if (paymentAccount.isHalCashAccount())
```
I guess the model needs some clean-up after copy & pasting :wink:

ripcurlx commented on desktop/src/main/java/bisq/desktop/main/offer/AtomicOfferView.java:
```diff
@@ -0,0 +1,753 @@
+            tradeFeeInBtcToggle.setManaged(false);
+            tradeFeeInBsqToggle.setVisible(false);
+            tradeFeeInBsqToggle.setManaged(false);
+        }
```
I guess all those trade fee UI components won't be necessary anymore for atomic BSQ trades.

ripcurlx commented on core/src/main/java/bisq/core/trade/Trade.java:
```diff
@@ -30,6 +30,7 @@
 import bisq.core.support.dispute.refund.RefundResultState;
 import bisq.core.support.messages.ChatMessage;
 import bisq.core.trade.protocol.ProcessModel;
+import bisq.core.trade.protocol.ProcessModelI;
```
Same here I find adding the I to the end to mark the interface kind of confusing. It could be mistaken as the short form of `Impl`.

chimp1984 commented:
I just looked mostly at the UX and looks good to me overall. Congrats!

Here some remarks:
I think the field in red should be hidden.
<img width="1066" alt="Screen Shot 2021-09-28 at 21 37 09" src="https://user-images.githubusercontent.com/54558767/135154443-7248821f-c106-43f4-abba-850b79b374d2.png">

Maybe better to call it "Atomic swaps" instead of "Atomic trades" 
Also in the take offer screen maybe better to use: SWAP instead of ATOMIC!!
And "BSQ swap" instead of "Atomic BSQ" in offerbook

In general maybe better not use `atomic` at all in the UI. For non techies it might sound confusion (nuclear - might create more environmental concerns ;-)). Maybe just call it BSQ swaps (swap might be too generic if we once add cross chain swaps)?

I got that expection after having one swap done and started another without creating a new block in between. I guess the BSQ change was not confirmed....
```
java.lang.IllegalArgumentException: Failed to prepare maker inputs
	at com.google.common.base.Preconditions.checkArgument(Preconditions.java:142)
	at bisq.core.trade.atomic.protocol.tasks.maker.AtomicMakerVerifiesTakerInputs.run(AtomicMakerVerifiesTakerInputs.java:58)
	at bisq.common.taskrunner.TaskRunner.next(TaskRunner.java:69)
	at bisq.common.taskrunner.TaskRunner.handleComplete(TaskRunner.java:85)
	at bisq.common.taskrunner.Task.complete(Task.java:58)
	at bisq.core.trade.protocol.tasks.AtomicTradeTask.complete(AtomicTradeTask.java:45)
	at bisq.core.trade.atomic.protocol.tasks.maker.AtomicMakerVerifiesAmounts.run(AtomicMakerVerifiesAmounts.java:85)
```
After creating a block it worked. Could take same offer.

In the code maybe better use BsqSwap instead of Atomic. AtomicProcessModel sounds confusing as the ProcessModel is not atomic but a ProcessModel for an atomic bsq swap.

Also would avoid the postfix I like in ProcessModelI.
Better rename the ProcessModel with some short prefix (DefaultProcessModel if no better found)

I think addP2PMessageFilter will not work as filterManager.getFilter() is likley null at startup. I would also use a concrete type for message not Object. And would invert the result as now false is expected that the predicate is evaluated true.

chimp1984 committed:
Rename OfferPayloadI to OfferPayloadBase

chimp1984 committed:
Fix var names ans strings using offerPayloadI

chimp1984 committed:
Add BSQ_SWAP_OFFER capability

We need to use capabilities to check if we can send a
msg to the peer containing BSQ_Swap offer data.
Added also support for BundleOfEnvelopes in the
noCapabilityRequiredOrCapabilityIsSupported method and
refactored the method.

chimp1984 committed:
Rename noCapabilityRequiredOrCapabilityIsSupported to testCapability

sqrrm committed:
Merge remote-tracking branch 'origin/master' into atomic-bsq-trades

chimp1984 committed:
Remove handling for bundledEnvelopes in send method

The code was never executed as it is not usual that on
one connection multiple messages are sent in a short period.
The handling for bundledEnvelopes has been implemented later
into the broadcast domain, there is does do its job.

chimp1984 committed:
Rename AtomicOfferPayload to BsqSwapOfferPayload

chimp1984 committed:
Add extraDataMap to BsqSwapOfferPayload

chimp1984 committed:
Rename AtomicAccountPayload to BsqSwapAccountPayload

Remove generic_string (was not used)

chimp1984 committed:
Rename AtomicAccount to BsqSwapAccount

chimp1984 committed:
Rename PaymentMethod.ATOMIC to PaymentMethod.BSQ_SWAP

chimp1984 committed:
Rename ProcessModelI to TradeProtocolModel

chimp1984 committed:
Rename AtomicProcessModel to BsqSwapProtocolModel

chimp1984 committed:
Rename AtomicProcessModel to BsqSwapProtocolModel in protobuf

chimp1984 committed:
Remove unused params

chimp1984 committed:
Rename AtomicTrade to BsqSwapTrade

chimp1984 committed:
Rename AtomicTrade to BsqSwapTrade in protobuf

chimp1984 committed:
Rename AtomicTradeManager to BsqSwapTradeManager

chimp1984 committed:
Move bisq.core.trade.atomic.AtomicTxBuilder to bisq.core.btc.AtomicTxBuilder

As it is about transaction domain I think it fits better in the bitcoin domain

chimp1984 committed:
Rename AtomicTxBuilder to BsqSwapTxHelper

Builder would suggest the builder pattern, so maybe
better use helper.

chimp1984 committed:
Rename AtomicMakerProtocol to BsqSwapMakerProtocol
and AtomicTakerProtocol to BsqSwapTakerProtocol

chimp1984 committed:
Move bisq.core.trade.protocol.BsqSwapProtocolModel to bisq.core.trade.atomic.protocol.BsqSwapProtocolModel

chimp1984 committed:
Move trade model classes to bisq.core.trade.model package

Keep bsqSwap and normal trade separate. Leave common classed
in base package.
No good name yet for default trade (default cannot be used as package name)

chimp1984 committed:
Move protocol classes to bisq.core.trade.protocol package

chimp1984 committed:
Move common classes level up

chimp1984 committed:
Rename ProcessModelServiceProvider to Provider

chimp1984 committed:
Move common classes to misc package

chimp1984 committed:
Move bisq.core.trade.protocol.messages to bisq.core.trade.messages

chimp1984 committed:
Move TradeManager, MakerTrade, TakerTrade one level up

Cleanups

chimp1984 committed:
Update protocol and protocol model classes

chimp1984 committed:
Remove Atomic prefixes for tasks

The package should be enough to qualify them as
BsqSwap tasks. They are only used in the limited scope
of the protocol package.

chimp1984 committed:
Rename CreateAtomicTxRequest and CreateAtomicTxResponse

chimp1984 committed:
Rename methods, fields with atomic prefix

chimp1984 committed:
Rename views

chimp1984 committed:
Rename AtomicOfferFunding

chimp1984 committed:
Refactor placeoffer packages

chimp1984 committed:
Replace TransactionResultHandler with Runnable

There is no tx set at bsqSwap place offer protocol

chimp1984 committed:
Replace ocurrances of atomic with bsqSwap

chimp1984 committed:
Use OfferPayloadBase instead of ProtectedStoragePayload as type

As we are in the offer domain we should use the domain relevant type.

chimp1984 committed:
Remove redundant OfferPayload.Direction

Move Direction one level up and rename to OfferDirection
to match protobuf definition

chimp1984 committed:
Move AvailabilityResult to bisq.core.offer.availability

chimp1984 committed:
Small refactorings, cleanups

chimp1984 committed:
Refactor BsqSwapWalletWatcher and OpenOfferManager

I think the dependency should be inverted here, so that the
OpenOfferManager does not know about the details about BsqSwap
funding.
Removed FundingListener, use listener on observable list instead
and maintain a filtered list in the BsqSwapWalletWatcher.

chimp1984 committed:
Add counter to fee service check

chimp1984 committed:
Fix copy paste error

sqrrm committed:
Merge pull request #9 from chimp1984/Bsq-swap-3

Bisq-swap-3

chimp1984 committed:
Add common interface for tradePeer.
Move TradingPeer to trade package.

Only stub now...

chimp1984 committed:
Remove payment method handling

chimp1984 committed:
Add generics for TradePeer

chimp1984 committed:
Rename getTradingPeer to getTradePeer

chimp1984 committed:
Replace TradingPeer BsqSwapTradePeer in BsqSwapProtocolModel

chimp1984 committed:
Add the 4 variants to trade and protocols

Comment out all existing code in protocol domain. Will get re-added step by step...

chimp1984 committed:
Trade protocol update according to other trade protocol

- Follow the buyer/seller and maker/taker perspectives
  according to the domain context
- Use the protocol model only for state data required as
  shared model for the tasks
- Add a tradePeer model for holding the peers data
- Keep logic in tasks and use a util class for re-used methods

chimp1984 committed:
Use tx confidence listener in SellerSetupTxListener

chimp1984 committed:
Cleanups

chimp1984 committed:
Rename bsqSwapProtocolModel to protocolModel

chimp1984 committed:
Rename bsqSwapTrade to trade

chimp1984 committed:
Rename TradeProtocolModel to ProtocolModel

chimp1984 committed:
Rename bsqSwapTrade to trade

chimp1984 committed:
Move package bisq.core.trade.messages to bisq.core.trade.protocol.messages

As messages are part of the protocol its better inside the protocol package

chimp1984 committed:
Move ClosedTradableManager and FailedTradesManager inside bisq.core.trade.model

chimp1984 committed:
Remove references from multisigbased trade domain from generic trade classes
Use interfaces for state and phase

chimp1984 committed:
Use concrete state/phase in impl classes
Cleanups

chimp1984 committed:
Move contract into trade package

chimp1984 committed:
Use Trade as type in tradableList
Make delegates to bsqSwapTradeManager more explicit
Cleanup stream concats
Add BsqTradeDetailsWindow instead using OfferDetailsWindow
(not impl. changes yet in window)

chimp1984 committed:
Remove offer only once trade is completed (tx visisble if maker is seller)
Pass bsqWalletService to getTransaction as provider is not set for
closed trades but we want to access it in views.

chimp1984 committed:
Rename classes

chimp1984 committed:
Rename trade package to bisq_v1

chimp1984 committed:
Move TradeProtocolFactory level up

chimp1984 committed:
Rename bsqswap package to bsq_swap

chimp1984 committed:
Start work on BsqTradeDetailsWindow

chimp1984 committed:
Add comment to make fee allocation in tx more clear

chimp1984 committed:
Rename BsqSwapWalletWatcher to BsqSwapOfferManager
Fix params in test

chimp1984 committed:
Add methods to DaoFacade and DaoStateService

chimp1984 committed:
Add fee listener

chimp1984 committed:
Use Singleton annotation

chimp1984 committed:
Let BsqSwapAccountPayload extend PaymentAccountPayload
We do not use an address but use that account only as
proxy for marking Bsq Swaps

chimp1984 committed:
Add comment

chimp1984 committed:
Increase fee tolerance.
Fee can be quite volatile so we allow larger deviations

chimp1984 committed:
Rename getTradeVolume to getVolume

chimp1984 committed:
Add getBsqTradeAmount to BsqSwapCalculation
Use that method in BsqTrade.getBsqTradeAmount()

chimp1984 committed:
Check for witness before setting it

chimp1984 committed:
Remove unnecessary check, add comment

chimp1984 committed:
Add todo

chimp1984 committed:
Add comment

chimp1984 committed:
Refactor tx creation and verification

Add message with finalized tx to be sent to seller.
Add Publish tx task to sellers side (only dont if
tx not already received)

chimp1984 committed:
Add handling for trade complete in case we have not got it set already

chimp1984 committed:
Extract super class for sending messages

chimp1984 committed:
Cleanups

chimp1984 committed:
Complete trade in case the tx broadcast got a timeout

chimp1984 committed:
Remove unused methods

chimp1984 committed:
Add scriptTypeId to RawTransactionInput

We will need that to know if its a segwit input or not.
Add constructors for converting a TransactionInput to
RawTransactionInput. Handle bitcoinSerialize according
to segwit flag.

This change should not break anything as we only add a
field to protobuf and RawTransactionInput is not used in
cases where we depend on the hash of the payload like in
offers.

chimp1984 committed:
Replace constructor call with fromProto

chimp1984 committed:
Remove unused methods

chimp1984 committed:
Use RawTransactionInput constructor instead of getRawInputFromTransactionInput

chimp1984 committed:
Move getRawInputFromTransactionInput to TradeWalletService

Don't change old behaviour for normal trade protocol methods
where we use segwit=false for serialisation.

chimp1984 committed:
Fix wrong exception, add more log details

chimp1984 committed:
Rename getTxSize to getVBytesSize

chimp1984 committed:
Fix check for segwit.
transactionInput.hasWitness would only work for signed
inputs but we need to do the fee calculation before the signing.

Adjust fee calculation values using vbytes
according to https://bitcoin.stackexchange.com/questions/87275/how-to-calculate-segwit-transaction-fee-in-bytes

chimp1984 committed:
Add handling for InsufficientMoneyException

This need to be tested further...

chimp1984 committed:
Add verification for peers change amount

chimp1984 committed:
Add verification for peers change

chimp1984 committed:
Store sellersBsqPayoutAmount in protocolModel.setPayout
at process input task and reuse that value in next task

chimp1984 committed:
Fix checks for change with trade fees and tx fees

chimp1984 committed:
Comment out balance check (temp for dev)

chimp1984 committed:
Do not handle dust here. We will handle it in caller

chimp1984 committed:
Fix handling for slightly overpaid BTC funds

If available funds is < dust it will add the dust change to miner fee
If the If available funds is > dust but the additional change output
would produce another dust change it will also add it to miner fees.
In that case the added amount is > dust.

chimp1984 committed:
Add address validation

chimp1984 committed:
Add check for change <= expectedChange

chimp1984 committed:
Update protocol description

chimp1984 committed:
Rename methods and params

chimp1984 committed:
Add getParentTxId method

Cleanup, make constructors private and
add deprecated annotation

chimp1984 committed:
Use getParentTxId method

chimp1984 committed:
Use isDust

chimp1984 committed:
Add getUnspentTxOutputValue method
Verify also sum of input values from DAO state data

chimp1984 committed:
Handle dust in caller
Add comments

sqrrm committed:
Merge pull request #16 from chimp1984/Complete-trade-protocol

Complete trade protocol

ripcurlx commented:
Closing as superseded and merged by #5775
